### PR TITLE
✨[story-ads] Introduce new placement algorithm

### PR DESCRIPTION
**Pull request by calebcordry**
Determines number of pages and insertion points upon page load. Monetizes shorter stories (>4 pages).

Partial #33147

### CODE REVIEW

File: extensions/amp-story-auto-ads/0.1/algorithm-utils.js [hunk 1]
```diff
@@ -13,15 +13,22 @@
...
 
 /**
  * Choose placement algorithm implementation.
+ * @param {!Window} win
  * @param {!StoryStoreService} storeService
  * @param {!StoryAdPageManager} pageManager
  * @return {!StoryAdPlacementAlgorithm}
  */
-export function getPlacementAlgo(storeService, pageManager) {
+export function getPlacementAlgo(win, storeService, pageManager) {
```
Comment (calebcordry) [inline extensions/amp-story-auto-ads/0.1/algorithm-utils.js:28]: Make sure to pass `win` in `amp-story-auto-ads.js` call site once #33343 is merged.

Comment (calebcordry) [reply extensions/amp-story-auto-ads/0.1/algorithm-utils.js:28]: Decided to do this in a new PR so we can have clean revert if necessary.

File: extensions/amp-story-auto-ads/0.1/algorithm-predetermined.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * Calculate the indices of where ads should be placed based
+ * on story length and the number of ads we want to show.
+ * @param {number} storyLength
+ * @param {number} numberOfAds
+ * @return {!Array<number>}
+ * @visibleForTesting
+ */
+export function getAdPositions(storyLength, numberOfAds) {
```
Comment (powerivq) [inline extensions/amp-story-auto-ads/0.1/algorithm-predetermined.js:36]: Q: Is `numberOfAds` guaranteed to be non-zero?

Comment (calebcordry) [reply extensions/amp-story-auto-ads/0.1/algorithm-predetermined.js:36]: Nope, and I was not handling it correctly. Added early exit and test case. Good catch.

**Review by powerivq — APPROVED** [review]

**Comment by rsimha** [comment]
@calebcordry This PR seems to have broken dependency checks: https://app.circleci.com/pipelines/github/ampproject/amphtml/5465/workflows/0d3541e6-3f34-49a1-bf1b-28fef021cece/jobs/74724/parallel-runs/0/steps/0-108

Perhaps a conflicting change snuck in after this was tested but before it was merged?

**Comment by calebcordry** [comment]
@rsimha thanks for the heads up. Will send fix shortly.

**Pull request merged**